syntax/zsh.vim

```diff
@@ -84,9 +84,13 @@ syn match   zshFunction         '^\s*\k\+\ze\s*()'
 
 syn match   zshOperator         '||\|&&\|;\|&!\='
 
+                                " <<<, <, <>, and variants.
 syn match   zshRedir            '\d\=\(<<<\|<&\s*[0-9p-]\=\|<>\?\)'
+                                " >, >>, and variants.
 syn match   zshRedir            '\d\=\(>&\s*[0-9p-]\=\|&>>\?\|>>\?&\?\)[|!]\='
-syn match   zshRedir            '|&\='
+                                " | and |&, but only if it's not preceeded or
+                                " followed by a | to avoid matching ||.
+syn match   zshRedir            '|\@<!|&\=|\@!'
```

Comment:
This regexp is a bit uglier/more involved than I'd like, but I couldn't really figure out a good way to do it otherwise.

Reply:
FWIW, even though it might not make a big difference for the performance, I would limit the backtracking to 1 byte:

    syn match   zshRedir            '|\@<!|&\=|\@!'
    syn match   zshRedir            '|\@1<!|&\=|\@!'
                                        ^

Often, when a syntax rule causes Vim to be slow, it's because of backtracking.  So, whenever it applies to a text with a fixed length, I compute the latter and specify a limit:

    \%(abc\)\@3<=
              ^

I seem to remember that it makes the rule a little faster (though not that much).

Reply:
Thanks; there's a small performance win:

	./testing.vim/tvim bench-syn ~cp/tpad.zsh/tpad.zsh:100
		TOTAL      COUNT  MATCH   SLOWEST     AVERAGE   NAME               PATTERN
		0.027522   2271   1       0.000079    0.000012  zshRedir           |\@<!|&\=|\@!
		0.027138   2271   1       0.000069    0.000012  zshRedir           |\@1<!|&\=|\@!

	./testing.vim/tvim bench-syn ./testdata/redir.zsh:20
		TOTAL      COUNT  MATCH   SLOWEST     AVERAGE   NAME               PATTERN
		0.018359   6800   3700    0.000030    0.000003  zshRedir           |\@<!|&\=|\@!
		0.018239   6800   3700    0.000037    0.000003  zshRedir           |\@1<!|&\=|\@!

That's about 0.0004s over 20 redraws so I don't think anyone will notice 🙃 But small changes can add up.